### PR TITLE
Networking refactor

### PR DESCRIPTION
**Pull request by gastonmaspero**
## Summary ##

**Networking Refactor**

Refactored API requests using Wolmo Networking

### iPhone X

Simulator Screen Shot - iPhone X - 2019-06-13 at 15 52 34
Simulator Screen Shot - iPhone X - 2019-06-13 at 15 52 44
Simulator Screen Shot - iPhone X - 2019-06-13 at 15 52 55
Simulator Screen Shot - iPhone X - 2019-06-13 at 15 55 43
rent
screencast

### CODE REVIEW

File: WBooks/AddNew/AddNewController.swift [hunk 1]
```diff
@@ -71,42 +71,56 @@ class AddNewController: UIViewController {
...
+            guard let this = self else {
+                return
+            }
+            this._loading = true
+            this._view.disableSubmit()
+            this._view.disableInteractions()
+            this._viewModel.addBook()
+        }
+        //Actions when submit request is completed
+        _viewModel.addState.producer.startWithValues { [weak self] state in
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift:103]: You can use reactive swift `Actions` to manage this kind of requests. With that you can observe values and errors and manage each separately without the enum. You don't need to change it in this training, but keep it in mind :D

File: WBooks/AddNew/AddNewViewModel.swift [hunk 2]
```diff
@@ -18,12 +24,27 @@ class AddNewViewModel {
...
     let descrpition = MutableProperty("")
     
+    let bookRepository = RepositoryBuilder.getDefaultBookRepository()
+    
+    let addState = MutableProperty(AddBookState.sleep)
+    
     private var book: Book {
-        return Book(title: bookName.value, author: author.value, id: 0, genre: genre.value,
-                    year: year.value, image: "some url", status: "available")
+        return Book(id: 0, title: bookName.value, author: author.value, genre: genre.value, year: year.value, image: "some url", status: "available")
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewViewModel.swift]: break the line

File: WBooks/Book Comments/BookCommentsViewModel.swift [hunk 3]
```diff
@@ -9,36 +9,31 @@
...
-        state.value = .error
-        return
+    func loadComments(bookID: Int) {
+        commentsRepository.fetchComments(bookID: bookID).startWithResult { [weak self] result in
+            switch result {
+            case .success(let resultArray):
+                self?.comments = resultArray
+                self?.state.value = resultArray.isEmpty ? .empty : .withValues
+            case .failure(let error):
+                // Here you can use error if you need it
```
Comment (carcos-wolox) [inline WBooks/Book Comments/BookCommentsViewModel.swift]: This was just for the guide. If you don't need it delete it

File: WBooks/Book Details/BookDetailsViewController.swift [hunk 4]
```diff
@@ -36,47 +36,56 @@ class BookDetailsViewController: UIViewController {
...
             self?.rent()
         }
+        
+        //Actions when rent request is completed
+        _viewModel.rentState.producer.startWithValues { [weak self] state in
+            guard let this = self else {
+                return
+            }
+            switch state {
+            case .rentError:
```
Comment (carcos-wolox) [inline WBooks/Book Details/BookDetailsViewController.swift]: this enum seems to be "equal" to add enum.. you could have only one generic enum if they have the same cases

File: WBooks/Book Details/BookDetailsViewController.swift [hunk 5]
```diff
@@ -36,47 +36,56 @@ class BookDetailsViewController: UIViewController {
...
-    
+
     func rent() {
-        switch _viewModel.status {
+        switch _viewModel.status.value {
         case .available:
-            let alert = UIAlertController(alertViewModel: ErrorAlertViewModel(title: "THANKS".localized(), message: "PROCESSING_RENT".localized(), dismissButtonTitle: "ACCEPT".localized()))
-            self.present(alert, animated: true, completion: nil)
-            _viewModel.rentBook(onSuccess: onBookRentSuccess, onError: onBookRentError)
+            present(_view.processingRentAlert, animated: true, completion: nil)
```
Comment (carcos-wolox) [inline WBooks/Book Details/BookDetailsViewController.swift]: Why did you set this alert in the view?

File: WBooks/Book Details/BookDetailsViewModel.swift [hunk 6]
```diff
@@ -42,45 +43,69 @@ enum Availability: String {
...
-        bookViewModel.status = "rented"
-        self.status = .inHands
+        
+        rentsRepository.rentBook(bookID: bookViewModel.id, from: today, to: tomorrow).startWithResult { [weak self] result in
+            guard let this = self else {
+                return
+            }
+            switch result {
+            case .success:
+                this.bookViewModel.status = "rented"
```
Comment (carcos-wolox) [inline WBooks/Book Details/BookDetailsViewModel.swift:93]: Localized

File: WBooks/Library/MainMenuController.swift [hunk 7]
```diff
@@ -39,15 +39,11 @@ class MainMenuController: UIViewController {
...
     }
     
-    func onSuccess(books: [Book]) {
-        _viewModel.onSuccess(books: books)
-    }
-    
     private func configureTableView() {
-        _viewModel.books.producer.startWithValues { [weak self] _ in
+        _viewModel.state.producer.startWithValues { [weak self] _ in
             self?._view.tableView.reloadData()
```
Comment (carcos-wolox) [inline WBooks/Library/MainMenuController.swift]: You are reloading the tableView each time `state` changes. Shouldn't be only when the array is not empty?

**Review by carcos-wolox — APPROVED** [review]
Check the comments before merging. 👍 It's good!

File: WBooks/AddNew/AddNewController.swift (hunk 1 above)
Comment (gastonmaspero) [reply WBooks/AddNew/AddNewController.swift:103]: Cool =D

File: WBooks/Book Details/BookDetailsViewController.swift (hunk 5 above)
Comment (gastonmaspero) [reply WBooks/Book Details/BookDetailsViewController.swift]: It better belongs to the controller now that you point this out. I'll change it.

File: WBooks/Book Details/BookDetailsViewModel.swift (hunk 6 above)
Comment (gastonmaspero) [reply WBooks/Book Details/BookDetailsViewModel.swift:93]: That is just a property of the bookViewModel. When I set the labels that correspond to this information, I use the enum Availability, which has a computed property called text, that returns the book status in a localized fashion.

File: WBooks/Library/MainMenuController.swift (hunk 7 above)
Comment (gastonmaspero) [reply WBooks/Library/MainMenuController.swift]: Indeed. I'll change it!

**Pull request merged**